File: moai-adk-ts/src/core/installer/backup-utils.ts
// @CODE:INIT-003:DATA | SPEC: SPEC-INIT-003.md
// Related: @CODE:INIT-003:BACKUP, @CODE:INIT-003:MERGE

/**
 * @file Backup Utility Functions (v0.2.1)
 * @author MoAI Team
 * @tags @CODE:INIT-003:DATA
 *
 * Common utilities for backup creation across Phase A (moai init)
 * and Phase B (/alfred:0-project emergency backup).
 */

import * as fs from 'node:fs';
import * as path from 'node:path';

/**
 * Check if any MoAI-ADK files exist in a directory (OR condition)
 * @CODE:INIT-003:DATA - v0.2.1 selective backup check
 *
 * @param projectPath - Project directory path
 * @returns True if at least one MoAI-ADK file exists
 */
export function hasAnyMoAIFiles(projectPath: string): boolean {
  const hasClaudeDir = fs.existsSync(path.join(projectPath, '.claude'));
  const hasMoaiDir = fs.existsSync(path.join(projectPath, '.moai'));
  const hasClaudeMd = fs.existsSync(path.join(projectPath, 'CLAUDE.md'));

  return hasClaudeDir || hasMoaiDir || hasClaudeMd;
}

/**
 * Generate backup directory name with timestamp
 * @CODE:INIT-003:DATA
 *
 * @returns Backup directory name (e.g., `.moai-backup-2025-10-07T05-04-03`)
 */
export function generateBackupDirName(): string {
  const timestamp = new Date().toISOString().replace(/[:.]/g, '-').slice(0, -5);
  return `.moai-backup-${timestamp}`;
}

/**
 * Get list of files that need to be backed up
 * @CODE:INIT-003:DATA - v0.2.1 selective backup
 *
 * @param projectPath - Project directory path
 * @returns Array of file/directory paths that should be backed up
 */
export function getBackupTargets(projectPath: string): string[] {
  const targets: string[] = [];

  // Check directories
  const criticalDirs = ['.claude', '.moai'];
  for (const dir of criticalDirs) {
    if (fs.existsSync(path.join(projectPath, dir))) {
      targets.push(dir);
    }
  }

  // Check files
  const criticalFiles = ['CLAUDE.md'];
  for (const file of criticalFiles) {
    if (fs.existsSync(path.join(projectPath, file))) {
      targets.push(file);
    }
  }

  return targets;
}

/**
 * Copy directory recursively (sync version for phase-executor)
 * @CODE:INIT-003:DATA
 *
 * @param src - Source directory
 * @param dest - Destination directory
 */
export async function copyDirectoryRecursive(
  src: string,
  dest: string
): Promise<void> {
  await fs.promises.mkdir(dest, { recursive: true });
  const entries = await fs.promises.readdir(src, { withFileTypes: true });

  for (const entry of entries) {
    const srcPath = path.join(src, entry.name);
    const destPath = path.join(dest, entry.name);

    if (entry.isDirectory()) {
      await copyDirectoryRecursive(srcPath, destPath);
    } else {
      await fs.promises.copyFile(srcPath, destPath);
    }
  }
}

/**
 * Validate backup metadata structure
 * @CODE:INIT-003:DATA
 *
 * @param metadata - Backup metadata object
 * @returns True if metadata is valid
 */
export function isValidBackupMetadata(metadata: unknown): boolean {
  if (!metadata || typeof metadata !== 'object') {
    return false;
  }

  const meta = metadata as Record<string, unknown>;

  return (
    typeof meta.timestamp === 'string' &&
    typeof meta.backup_path === 'string' &&
    Array.isArray(meta.backed_up_files) &&
    typeof meta.status === 'string' &&
    ['pending', 'merged', 'ignored'].includes(meta.status as string) &&
    typeof meta.created_by === 'string'
  );
}

File: src/moai_adk/core/project/backup_utils.py
# @CODE:INIT-003:BACKUP | SPEC: .moai/specs/SPEC-INIT-003/spec.md | TEST: tests/unit/test_backup_utils.py
"""백업 유틸리티 모듈 (SPEC-INIT-003 v0.3.0)

Selective Backup 전략:
- 필요한 파일만 백업 (OR 조건)
- 백업 경로: .moai/backups/{timestamp}/ (v0.3.0)
"""

from datetime import datetime
from pathlib import Path


# 백업 대상 파일/디렉토리 (OR 조건 - 하나라도 있으면 백업)
BACKUP_TARGETS = [
    ".moai/config.json",
    ".moai/project/",
    ".moai/memory/",
    ".claude/",
    "CLAUDE.md",
]

# 사용자 데이터 보호 경로 (백업에서 제외)
PROTECTED_PATHS = [
    ".moai/specs/",
    ".moai/reports/",
]


def has_any_moai_files(project_path: Path) -> bool:
    """MoAI-ADK 파일 존재 여부 확인 (OR 조건)

    Args:
        project_path: 프로젝트 경로

    Returns:
        하나라도 존재하면 True
    """
    for target in BACKUP_TARGETS:
        target_path = project_path / target
        if target_path.exists():
            return True
    return False


def get_backup_targets(project_path: Path) -> list[str]:
    """백업 대상 파일/디렉토리 목록 반환

    Args:
        project_path: 프로젝트 경로

    Returns:
        존재하는 백업 대상 목록
    """
    targets: list[str] = []
    for target in BACKUP_TARGETS:
        target_path = project_path / target
        if target_path.exists():
            targets.append(target)
    return targets


def generate_backup_dir_name() -> str:
    """타임스탬프 기반 백업 디렉토리명 생성 (v0.3.0)

    Returns:
        YYYYMMDD-HHMMSS 형식 (순수 타임스탬프)
        Note: .moai/backups/ 경로는 호출자가 추가
    """
    timestamp = datetime.now().strftime("%Y%m%d-%H%M%S")
    return timestamp


def is_protected_path(rel_path: Path) -> bool:
    """보호 경로 여부 확인

    Args:
        rel_path: 상대 경로

    Returns:
        보호 경로이면 True
    """
    rel_str = str(rel_path).replace("\\", "/")
    return any(
        rel_str.startswith(p.lstrip("./").rstrip("/")) for p in PROTECTED_PATHS
    )
